File: app/src/main/java/com/sd/demo/vtrack/MainActivity.java
package com.sd.demo.vtrack;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.sd.demo.vtrack.databinding.ActivityMainBinding;
import com.sd.lib.vtrack.ext.FPositionTracker;
import com.sd.lib.vtrack.tracker.ViewTracker;
import com.sd.lib.vtrack.tracker.ViewTracker.Position;
import com.sd.lib.vtrack.updater.ViewUpdater;
import com.sd.lib.vtrack.updater.impl.OnPreDrawUpdater;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    public static final String TAG = MainActivity.class.getSimpleName();

    private ActivityMainBinding mBinding;
    private FPositionTracker mViewTracker;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());

        getViewTracker().setSource(mBinding.viewSource);
        getViewTracker().setTarget(mBinding.viewTarget);
    }

    private FPositionTracker getViewTracker() {
        if (mViewTracker == null) {
            mViewTracker = new FPositionTracker() {
                @NonNull
                @Override
                protected ViewUpdater createTargetUpdater() {
                    return new OnPreDrawUpdater();
                }
            };
            // 设置回调对象
            mViewTracker.setCallback(new ViewTracker.Callback() {
                /**
                 * 按照指定的位置{@link Position}追踪到target后回调，回调source相对于父布局的x和y值
                 *
                 * @param x      source相对于父布局的x值
                 * @param y      source相对于父布局的y值
                 * @param source 源view
                 * @param target 目标view
                 */
                @Override
                public void onUpdate(int x, int y, @NonNull View source, @NonNull View target) {
                    Log.i(TAG, x + "," + y);
                    source.layout(x, y, x + source.getMeasuredWidth(), y + source.getMeasuredHeight());
                }
            });
            // 设置源view
            mViewTracker.setSource(mBinding.viewSource);
            // 设置目标view
            mViewTracker.setTarget(mBinding.viewTarget);
        }
        return mViewTracker;
    }

    @Override
    public void onClick(View v) {
        if (v == mBinding.btnStart) {
            // 开始追踪
            getViewTracker().start();
        } else if (v == mBinding.btnStop) {
            // 停止追踪
            getViewTracker().stop();
        }

        if (v == mBinding.btnTopLeft) {
            getViewTracker().setPosition(Position.TopLeft);
        } else if (v == mBinding.btnTopCenter) {
            getViewTracker().setPosition(Position.TopCenter);
        } else if (v == mBinding.btnTopRight) {
            getViewTracker().setPosition(Position.TopRight);
        } else if (v == mBinding.btnLeftCenter) {
            getViewTracker().setPosition(Position.LeftCenter);
        } else if (v == mBinding.btnCenter) {
            getViewTracker().setPosition(Position.Center);
        } else if (v == mBinding.btnRightCenter) {
            getViewTracker().setPosition(Position.RightCenter);
        } else if (v == mBinding.btnBottomLeft) {
            getViewTracker().setPosition(Position.BottomLeft);
        } else if (v == mBinding.btnBottomCenter) {
            getViewTracker().setPosition(Position.BottomCenter);
        } else if (v == mBinding.btnBottomRight) {
            getViewTracker().setPosition(Position.BottomRight);
        }
    }
}

File: app/src/main/java/com/sd/demo/vtrack/MainActivity.kt
package com.sd.demo.vtrack

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.sd.demo.vtrack.databinding.ActivityMainBinding
import com.sd.lib.vtrack.ext.FPositionTracker
import com.sd.lib.vtrack.tracker.ViewTracker
import com.sd.lib.vtrack.updater.ViewUpdater
import com.sd.lib.vtrack.updater.impl.OnPreDrawUpdater

class MainActivity : AppCompatActivity(), View.OnClickListener {
    private val _binding by lazy { ActivityMainBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(_binding.root)
    }

    private val tracker: FPositionTracker by lazy {
        object : FPositionTracker() {
            override fun createTargetUpdater(): ViewUpdater {
                return OnPreDrawUpdater()
            }
        }.apply {
            // 设置源view
            setSource(_binding.viewSource)
            // 设置目标view
            setTarget(_binding.viewTarget)
            // 设置回调对象
            setCallback(object : ViewTracker.Callback() {
                /**
                 * 按照指定的位置[Position]追踪到target后回调，回调source相对于父布局的x和y值
                 *
                 * @param x      source相对于父布局的x值
                 * @param y      source相对于父布局的y值
                 * @param source 源view
                 * @param target 目标view
                 */
                override fun onUpdate(x: Int, y: Int, source: View, target: View) {
                    Log.i(TAG, "$x,$y")
                    source.layout(x, y, x + source.measuredWidth, y + source.measuredHeight)
                }
            })
        }
    }

    override fun onClick(v: View) {
        when (v) {
            _binding.btnStart -> {
                // 开始追踪
                tracker.start()
            }
            _binding.btnStop -> {
                // 停止追踪
                tracker.stop()
            }

            _binding.btnTopLeft -> tracker.setPosition(ViewTracker.Position.TopLeft)
            _binding.btnTopCenter -> tracker.setPosition(ViewTracker.Position.TopCenter)
            _binding.btnTopRight -> tracker.setPosition(ViewTracker.Position.TopRight)

            _binding.btnLeftCenter -> tracker.setPosition(ViewTracker.Position.LeftCenter)
            _binding.btnCenter -> tracker.setPosition(ViewTracker.Position.Center)
            _binding.btnRightCenter -> tracker.setPosition(ViewTracker.Position.RightCenter)

            _binding.btnBottomLeft -> tracker.setPosition(ViewTracker.Position.BottomLeft)
            _binding.btnBottomCenter -> tracker.setPosition(ViewTracker.Position.BottomCenter)
            _binding.btnBottomRight -> tracker.setPosition(ViewTracker.Position.BottomRight)
        }
    }

    companion object {
        const val TAG = "MainActivity"
    }
}